### ISSUE TITLE
Why there is no '[CLS]' or '[EOS]' token added in my tokenizer?

### ISSUE THREAD

**Issue by Luoyang144** [huggingface/tokenizers#1208]
Hello, I' m tring to train a custom tokenizer from my corpus, but I find there is no [CLS] or [EOS] token added to input_ids, how to solve this problem?
Here is my training code:
```
from tokenizers import Tokenizer, models, normalizers, pre_tokenizers, trainers, decoders
from transformers import BertTokenizerFast

tokenizer = Tokenizer(models.WordLevel())

tokenizer.normalizer = normalizers.Sequence([
    normalizers.NFKC(),
    normalizers.Lowercase()
])

tokenizer.pre_tokenizer = pre_tokenizers.BertPreTokenizer()

special_tokens = ["[UNK]", "[PAD]", "[CLS]", '[EOS]', "[SEP]", "[MASK]"]
trainer = trainers.WordLevelTrainer(vocab_size=3500, special_tokens=special_tokens)


dataset = open("../data/input/corpus.txt")
batch_size = 10

def batch_iterator():
    result = []
    for l in dataset:
        result.append(l)
        if len(result) == batch_size:
            yield result
            result = []

tokenizer.train_from_iterator(batch_iterator(), trainer=trainer)

new_tokenizer = BertTokenizerFast(tokenizer_object=tokenizer)

new_tokenizer.save_pretrained("./tokenizer/")
```
And my test code:
```
from transformers import AutoTokenizer, BertTokenizer

example = ["22 12 38 41 17 81 10", "22 12 38 41 17 81 10 111"]
tokenizer = AutoTokenizer.from_pretrained("tokenizer/")
# print(tokenizer)
print(tokenizer(example, padding=True, add_special_tokens=True))
# print(tokenizer.tokenize(example))

tokenizer = BertTokenizer.from_pretrained('fnlp/bart-large-chinese')
# print(tokenizer.special_tokens_map)
print(tokenizer(example, padding=True))
# print(tokenizer.tokenize(example))
```
Results:
```
{'input_ids': [[38, 28, 54, 58, 33, 104, 26, 1], [38, 28, 54, 58, 33, 104, 26, 126]], 'token_type_ids': [[0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0]], 'attention_mask': [[1, 1, 1, 1, 1, 1, 1, 0], [1, 1, 1, 1, 1, 1, 1, 1]]}
{'input_ids': [[101, 30937, 30863, 31113, 31239, 30902, 31843, 30847, 102, 0], [101, 30937, 30863, 31113, 31239, 30902, 31843, 30847, 33142, 102]], 'token_type_ids': [[0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], 'attention_mask': [[1, 1, 1, 1, 1, 1, 1, 1, 1, 0], [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]}
```
As you can see, only padding token is added correctly, how can I automatically add other special tokens?

**Comment by Narsil** [comment]
I think you're looking for a `post_processor`. The most general being `TemplateProcessor`:

https://huggingface.co/docs/tokenizers/api/post-processors

Does that solve your issue ?

**Comment by Luoyang144** [comment]
@Narsil Thank you it works!
Here is my code(for those troubled by this problem):
```
from tokenizers import Tokenizer, models, normalizers, pre_tokenizers, trainers, decoders
from transformers import BertTokenizerFast
from tokenizers.processors import BertProcessing
tokenizer = Tokenizer(models.WordLevel())

tokenizer.normalizer = normalizers.Sequence([
    normalizers.NFKC(),
    normalizers.Lowercase()
])

tokenizer.pre_tokenizer = pre_tokenizers.BertPreTokenizer()
# tokenizer.decoder = decoders.ByteLevel()


special_tokens = ["[UNK]", "[PAD]", "[CLS]", '[EOS]', "[SEP]", "[MASK]"]
trainer = trainers.WordLevelTrainer(vocab_size=3500, special_tokens=special_tokens)


dataset = open("../data/input/corpus.txt")
batch_size = 10

def batch_iterator():
    result = []
    for l in dataset:
        result.append(l)
        if len(result) == batch_size:
            yield result
            result = []

tokenizer.train_from_iterator(batch_iterator(), trainer=trainer)

tokenizer.post_processor = BertProcessing(
        ("[SEP]", tokenizer.token_to_id("[SEP]")),
        ("[CLS]", tokenizer.token_to_id("[CLS]")),
    )


new_tokenizer = BertTokenizerFast(tokenizer_object=tokenizer)

new_tokenizer.save_pretrained("./tokenizer/")
```
Tokenize a example:
```
from transformers import AutoTokenizer, BertTokenizer, BertTokenizerFast

example = ["22 12 38 41 17 81 10", "22 12 38 41 17 81 10 111"]
tokenizer = AutoTokenizer.from_pretrained("utils/tokenizer/")
# print(tokenizer)
print(tokenizer(example, padding=True, add_special_tokens=True))
# print(tokenizer.tokenize(example))

tokenizer = BertTokenizer.from_pretrained('fnlp/bart-large-chinese')
# print(tokenizer.special_tokens_map)
print(tokenizer(example, padding=True))
# print(tokenizer.tokenize(example))
```
Results:
```
{'input_ids': [[2, 19, 8, 38, 43, 16, 117, 6, 4, 1], [2, 19, 8, 38, 43, 16, 117, 6, 99, 4]], 'token_type_ids': [[0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], 'attention_mask': [[1, 1, 1, 1, 1, 1, 1, 1, 1, 0], [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]}
{'input_ids': [[101, 30937, 30863, 31113, 31239, 30902, 31843, 30847, 102, 0], [101, 30937, 30863, 31113, 31239, 30902, 31843, 30847, 33142, 102]], 'token_type_ids': [[0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], 'attention_mask': [[1, 1, 1, 1, 1, 1, 1, 1, 1, 0], [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]}
```

**Comment by Narsil** [comment]
Marking as closed. Feel free to reopen
